From 153c358c4440c82c5bf67d44168ba408a91bb7ef Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Tue, 18 Aug 2015 17:13:54 +0300
Subject: [PATCH 0330/1240] pcie: dw: configure the root complex as PCIe bridge

to allow u-boot to see device beyond our root complex we
set the root complex class as pci-pci bridge allowing u-boot
to configure it and perform configuration reads over the bridge

Change-Id: I6e071de510bf0e84106e2fb4281b15ce25413ebf
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/22919
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 drivers/pci/pcie_dw.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/pci/pcie_dw.c b/drivers/pci/pcie_dw.c
index 27fb292..919a98e 100644
--- a/drivers/pci/pcie_dw.c
+++ b/drivers/pci/pcie_dw.c
@@ -177,6 +177,7 @@ int dw_pcie_init(int host_id, uintptr_t regs_base, struct pcie_win *mem_win,
 {
 	struct pci_controller *hose = &pci_hose[host_id];
 	uintptr_t temp;
+	u32 reg;
 
 	dw_pcie_regions_setup(regs_base, cfg_win);
 
@@ -199,6 +200,17 @@ int dw_pcie_init(int host_id, uintptr_t regs_base, struct pcie_win *mem_win,
 	hose->priv_data = (void *)regs_base;
 	hose->first_busno = first_busno;
 
+	/* CMD reg:I/O space, MEM space, and Bus Master Enable */
+	reg = readl(regs_base + PCI_COMMAND);
+	reg |= (PCI_COMMAND_IO | PCI_COMMAND_MEMORY | PCI_COMMAND_MASTER);
+	writel(reg, regs_base + PCI_COMMAND);
+
+	/* Set the CLASS_REV of RC CFG header to PCI_CLASS_BRIDGE_PCI */
+	reg = readl(regs_base + PCI_CLASS_REVISION);
+	reg &= ~(0xFFFF << 16);
+	reg |= (PCI_CLASS_BRIDGE_PCI << 16);
+	writel(reg, regs_base + PCI_CLASS_REVISION);
+
 	/* Register the host */
 	pci_register_hose(hose);
 
-- 
1.9.1

